java/src/main/native/src/row_conversion.cu

```diff
@@ -0,0 +1,636 @@
+                }
+            } // end column loop
+        } // end row copy
+        // wait for the row_group to be totally copied into shared memory
+        __syncthreads();
+
+        // Step 2: Copy the data back out
+        // We know row_size is always aligned with and a multiple of int64_t;
+        int64_t * long_shared = reinterpret_cast<int64_t *>(shared_data);
+        int64_t * long_output = reinterpret_cast<int64_t *>(output_data);
+
+        cudf::size_type shared_input_index = threadIdx.x + (threadIdx.y * blockDim.x);
+        cudf::size_type shared_input_stride = blockDim.x * blockDim.y;
+        cudf::size_type row_index_end = ((row_group_index + 1) * rows_per_group);
+        if (row_index_end > num_rows) {
+            row_index_end = num_rows;
+        }
+        cudf::size_type num_rows_in_group = row_index_end - (row_group_index * rows_per_group);
+        cudf::size_type shared_length = row_size * num_rows_in_group;
+
+        cudf::size_type shared_input_end = shared_length/sizeof(int64_t);
+
+        cudf::size_type start_output_index = (row_size * row_group_index * rows_per_group)/sizeof(int64_t);
+
+        for (cudf::size_type shared_index = shared_input_index; 
+                shared_index < shared_input_end;
+                shared_index += shared_input_stride) {
+            long_output[start_output_index + shared_index] = long_shared[shared_index];
+        }
+        __syncthreads();
+        // Go for the next round
+    }
+}
+
+/**
+ * Calculate the dimensions of the kernel for fixed width only columns.
+ * @param [in] num_columns the number of columns being copied.
+ * @param [in] num_rows the number of rows being copied.
+ * @param [in] size_per_row the size each row takes up when padded.
+ * @param [out] blocks the size of the blocks for the kernel
+ * @param [out] threads the size of the threads for the kernel
+ * @return the size in bytes of shared memory needed for each block.
+ */
+static int calc_fixed_width_kernel_dims(
+    const cudf::size_type num_columns,
+    const cudf::size_type num_rows,
+    const cudf::size_type size_per_row,
+    dim3 & blocks,
+    dim3 & threads) {
+
+    // We have found speed degrades when a thread handles more than 4 columns.
+    // Each block is 2 dimensional. The y dimension indicates the columns.
+    // We limit this to 32 threads in the y dimension so we can still
+    // have at least 32 threads in the x dimension (1 warp) which should
+    // result in better coalescing of memory operations. We also
+    // want to guarantee that we are processing a multiple of 32 threads
+    // in the x dimension because we use atomic operations at the block
+    // level when writing validity data out to main memory, and that would
+    // need to change if we split a word of validity data between blocks.
+    int y_block_size = (num_columns + 3) / 4;
+    if (y_block_size > 32) {
+      y_block_size = 32;
```

Comment:
I might have misunderstood, but Line#204 indicates:
```
    // We do not support copying a subset of the columns in a row yet, so we don't
    // currently support a row that is wider than shared memory
```
But we're also capping the block's Y-dimension at 32. Should we add `CUDF_EXPECTS()` to bail out if the number of columns in the table exceeds 32?

Reply:
Sorry I should update the comments here. Each Y dimension ideally handles about 4 columns.  This is what I found from tests on a V100 to be ideal.  No idea if it holds for other architectures too.  So we first try to get a y dimension where each thread is going to read/write about 4 columns.  If that fails then we let the number of columns increase above 4.  So we can handle an arbitrary number of columns.  What we cannot handle is a row that is larger than fits in shared memory.

For performance reasons we go through shared memory as we transform between rows/columns so all of the main memory accesses are coalesced. We also want to have at most 48KB of shared memory.  That limit varies some what between different compute capabilities but 48KB is the smallest amount supported by what we support so I went with that.  Also to take full advantage of coalescing reads/writes I want to make sure that the X dimension is always a multiple of 32.  This means that each row can be at most 48/32 or 1.5KB.

That is enough for 1365 byte columns or 722 short columns or 372 int columns or 189 long columns or depending on padding/etc some odd combination of all of them.  There is a check further down for this.